core/src/main/java/io/undertow/server/handlers/BlockingTimeoutHandler.java

```diff
@@ -0,0 +1,422 @@
+        }
+        next.handleRequest(exchange);
+    }
+
+    private static final class TimeoutStreamSourceConduit implements StreamSourceConduit {
+        private final StreamSourceConduit delegate;
+        private final ServerConnection serverConnection;
+        private final long timeoutNanos;
+        private long remaining;
+
+        TimeoutStreamSourceConduit(
+                StreamSourceConduit delegate,
+                ServerConnection serverConnection,
+                long timeoutNanos) {
+            this.delegate = delegate;
+            this.serverConnection = serverConnection;
+            this.timeoutNanos = timeoutNanos;
+            this.remaining = timeoutNanos;
+        }
+
+        @Override
+        public long transferTo(long position, long count, FileChannel fileChannel) throws IOException {
+            return resetTimeoutIfReadSucceeded(delegate.transferTo(position, count, fileChannel));
+        }
+
+        @Override
+        public long transferTo(long count, ByteBuffer byteBuffer, StreamSinkChannel streamSinkChannel) throws IOException {
+            return resetTimeoutIfReadSucceeded(delegate.transferTo(count, byteBuffer, streamSinkChannel));
+        }
+
+        @Override
+        public int read(ByteBuffer byteBuffer) throws IOException {
+            return resetTimeoutIfReadSucceeded(delegate.read(byteBuffer));
+        }
+
+        @Override
+        public long read(ByteBuffer[] byteBuffers, int offset, int length) throws IOException {
+            return resetTimeoutIfReadSucceeded(delegate.read(byteBuffers, offset, length));
+        }
+
+        @Override
+        public void terminateReads() throws IOException {
+            delegate.terminateReads();
+        }
+
+        @Override
+        public boolean isReadShutdown() {
+            return delegate.isReadShutdown();
+        }
+
+        @Override
+        public void resumeReads() {
+            delegate.resumeReads();
+        }
+
+        @Override
+        public void suspendReads() {
+            delegate.suspendReads();
+        }
+
+        @Override
+        public void wakeupReads() {
+            delegate.wakeupReads();
+        }
+
+        @Override
+        public boolean isReadResumed() {
+            return delegate.isReadResumed();
+        }
+
+        @Override
+        public void awaitReadable() throws IOException {
+            awaitReadable(remaining, TimeUnit.NANOSECONDS);
+        }
+
+        @Override
+        public void awaitReadable(long duration, TimeUnit unit) throws IOException {
+            long startTime = System.nanoTime();
+            long requestedNanos = unit.toNanos(duration);
+            try {
+                delegate.awaitReadable(Math.min(requestedNanos, remaining), TimeUnit.NANOSECONDS);
+            } finally {
+                remaining -= (System.nanoTime() - startTime);
+            }
+            if (remaining < 0) {
+                IoUtils.safeClose(serverConnection);
```

Comment:
Might be helpful to log something at debug level here and in the write timeout conduit when the timeout is reached.